### DISCUSSION TITLE
Configuration Nodes

### DISCUSSION THREAD

**Discussion by mtoko** [phoddie/node-red-mcu#99] [General]
Firstly, thanks you to Peter and Ralph for all the hard work and fantastic progress
I am using an MQTT node with credentials and I vaguely remember that the configuration node credentials are only sent on the first deploy, so before the node mcu plugin the procedure was to copy the flow into node red mcu  before deploy so that these were downloaded with the flow to the mcu. I assume now that I have the plugin working that the config node credentials would be downloaded.
Summary
I have a simple flow of an inject node and an MQTT out node to a MQTT broker that is known to work. The flow works correctly if there are no credentials in the MQTT out or the broker but does not connect if the node red mcu flow has credentials. I have tried to input credentials and then build before deploy (savings config setup) and also to build after deploy but neither works
The mcu is a ESP32/nodemcu
Any advice appreciated

**Comment by ralphwetzel** [comment]
Could you please share a flow that shows the issue you have ... with redacted / simulated credentials of course.

**Reply by mtoko** [reply]
This is the flow before deploy, you can see the credentials in the broker
`[
    {
        "id": "9c629d06eeea7a09",
        "type": "tab",
        "label": "Flow 1",
        "disabled": false,
        "info": "",
        "env": [],
        "_mcu": {
            "mcu": true
        }
    },
    {
        "id": "e42e258e6f3d5c8c",
        "type": "inject",
        "z": "9c629d06eeea7a09",
        "name": "",
        "props": [
            {
                "p": "payload"
            },
            {
                "p": "topic",
                "vt": "str"
            }
        ],
        "repeat": "20",
        "crontab": "",
        "once": false,
        "onceDelay": "5",
        "topic": "",
        "payload": "",
        "payloadType": "date",
        "_mcu": {
            "mcu": true
        },
        "x": 170,
        "y": 80,
        "wires": [
            [
                "da80f24bceaa4e51",
                "2b5ec21c66d9da1b"
            ]
        ]
    },
    {
        "id": "da80f24bceaa4e51",
        "type": "mqtt out",
        "z": "9c629d06eeea7a09",
        "name": "",
        "topic": "Testy76",
        "qos": "0",
        "retain": "false",
        "respTopic": "",
        "contentType": "",
        "userProps": "",
        "correl": "",
        "expiry": "",
        "broker": "9a98af129a0f5b68",
        "_mcu": {
            "mcu": true
        },
        "x": 520,
        "y": 40,
        "wires": []
    },
    {
        "id": "2b5ec21c66d9da1b",
        "type": "debug",
        "z": "9c629d06eeea7a09",
        "name": "debug 1",
        "active": false,
        "tosidebar": true,
        "console": false,
        "tostatus": false,
        "complete": "false",
        "statusVal": "",
        "statusType": "auto",
        "_mcu": {
            "mcu": true
        },
        "x": 500,
        "y": 160,
        "wires": []
    },
    {
        "id": "9a98af129a0f5b68",
        "type": "mqtt-broker",
        "name": "",
        "broker": "192.168.1.107",
        "port": "1883",
        "clientid": "",
        "autoConnect": true,
        "usetls": false,
        "protocolVersion": "4",
        "keepalive": "60",
        "cleansession": true,
        "birthTopic": "",
        "birthQos": "0",
        "birthPayload": "",
        "birthMsg": {},
        "closeTopic": "",
        "closeQos": "0",
        "closePayload": "",
        "closeMsg": {},
        "willTopic": "",
        "willQos": "0",
        "willPayload": "",
        "willMsg": {},
        "userProps": "",
        "sessionExpiry": "",
        "_mcu": {
            "mcu": false
        },
        "credentials": {
            "user": "user",
            "password": "password"
        }
    }
]`
And this is the flow after deploy, no credentials
`[
    {
        "id": "9c629d06eeea7a09",
        "type": "tab",
        "label": "Flow 1",
        "disabled": false,
        "info": "",
        "env": [],
        "_mcu": {
            "mcu": true
        }
    },
    {
        "id": "e42e258e6f3d5c8c",
        "type": "inject",
        "z": "9c629d06eeea7a09",
        "name": "",
        "props": [
            {
                "p": "payload"
            },
            {
                "p": "topic",
                "vt": "str"
            }
        ],
        "repeat": "20",
        "crontab": "",
        "once": false,
        "onceDelay": "5",
        "topic": "",
        "payload": "",
        "payloadType": "date",
        "_mcu": {
            "mcu": true
        },
        "x": 170,
        "y": 80,
        "wires": [
            [
                "da80f24bceaa4e51",
                "2b5ec21c66d9da1b"
            ]
        ]
    },
    {
        "id": "da80f24bceaa4e51",
        "type": "mqtt out",
        "z": "9c629d06eeea7a09",
        "name": "",
        "topic": "Testy76",
        "qos": "0",
        "retain": "false",
        "respTopic": "",
        "contentType": "",
        "userProps": "",
        "correl": "",
        "expiry": "",
        "broker": "9a98af129a0f5b68",
        "_mcu": {
            "mcu": true
        },
        "x": 520,
        "y": 40,
        "wires": []
    },
    {
        "id": "2b5ec21c66d9da1b",
        "type": "debug",
        "z": "9c629d06eeea7a09",
        "name": "debug 1",
        "active": false,
        "tosidebar": true,
        "console": false,
        "tostatus": false,
        "complete": "false",
        "statusVal": "",
        "statusType": "auto",
        "_mcu": {
            "mcu": true
        },
        "x": 500,
        "y": 160,
        "wires": []
    },
    {
        "id": "9a98af129a0f5b68",
        "type": "mqtt-broker",
        "name": "",
        "broker": "192.168.1.107",
        "port": "1883",
        "clientid": "",
        "autoConnect": true,
        "usetls": false,
        "protocolVersion": "4",
        "keepalive": "60",
        "cleansession": true,
        "birthTopic": "",
        "birthQos": "0",
        "birthPayload": "",
        "birthMsg": {},
        "closeTopic": "",
        "closeQos": "0",
        "closePayload": "",
        "closeMsg": {},
        "willTopic": "",
        "willQos": "0",
        "willPayload": "",
        "willMsg": {},
        "userProps": "",
        "sessionExpiry": "",
        "_mcu": {
            "mcu": false
        }
    }
]`

**Reply by ralphwetzel** [reply]
Thank you for this example.
As I just learned from [this documentation](https://nodered.org/docs/creating-nodes/credentials), credentials are handled separately from the standard flow data.
Thus neither the plugin nor node-red-mcu (to my knowledge) are currently prepared to manage or process credentials.
I propose, you raise an issue (in either repository) and we'll find a way to implement that.

**Reply by mtoko** [reply]
Thanks Ralph. I have raised the issue in node-red-mcu-plugin.

**Comment by phoddie** [comment]
@mtoko – The MQTT node implementations on the MCU will use the user and password credentials when present. However, as you note, Node-RED does not export this information for security reasons. You can enter them manually in the flow JSON. The MCU plug-in is in a position to access the credentials, so it could add them to the exported flow with some additional work.

**Reply by mtoko** [reply]
@phoddie Thanks for this advice. That worked! Just for anyone interested
1.) Complete your flow with MQTT nodes and configure the MQTT broker as normal, then export the flow BEFORE deploy this will preserve the broker credentials in the JSON. Paste this flow to a text editor (Notepad etc)
2.)  Deploy the flow
3.) Close the editor
4.) Open flows.json  in ~/.node-red 
5.) Find the MQTT broker node and add the credentials section at the bottom
6.) Stop and Restart Node Red
7.) Open the Node Red editor and build the flow using mcu plug in
This procedure only works ONCE, if you make changes to your flow and build again then the credentials need to be added again. It appears that each time you build the credentials are lost, I think?
See my example flows above with and without credentials

**Reply by phoddie** [reply]
> Thanks for this advice. That worked!

Cool! Thank for your confirming that and writing out the details for others.

> It appears that each time you build the credentials are lost, I think?

Right. The flows.json file is overwritten each time you build, so the edits are lost.

I would like to see a way for this to work automatically when possible. That's going to require cooperating from the Node-RED editing environment. @ralphwetzel has some ideas there, so perhaps it can be achieved.

**Reply by ralphwetzel** [reply]
> @ralphwetzel has some ideas there, so perhaps it can be achieved.

It is definitely achievable - and the code is already done.
The only / last element that has to be aligned is the "How".

How shall the credentials being made available for the `node-red-mcu` runtime or the `nodered2mcu` preprocessor?
I've some concerns (already stated [here](https://github.com/ralphwetzel/node-red-mcu-plugin/issues/28#issuecomment-1455181667)) putting them into `flows.json` - as this creates the risk people (unaware of this fact) copy / paste them somewhere.

There are other options (most probably even more than I discussed), and - as either the preprocessor or the runtime have to deal with this - the final call is on @phoddie to decide which way to go.

**Reply by mtoko** [reply]
Just another quick tip to add to the steps above. I have done all my flow testing using an MQTT broker with no credentials (on another Node Red instance and the MQTT broker node). This is quick and easy because the nodes connect without adding credentials each time. When I have a "final" working flow, and want to use a cloud broker, I build this and add the credentials in the "finished" version
